### PR TITLE
Midi 110/T5-training

### PR DESCRIPTION
**Pull request by WojciechMat**
### Training utils

The main objective of this pull request was to prepare training loops compatible with huggingface models and our datasets.

#### Code snippet

To train a model, use:

```python
from hydra import initialize, compose

from utils import vocab_size
from training_utils import train_model
from data.dataset import MyTokenizedMidiDataset
from data.tokenizer import MultiVelocityEncoder

# Get the config without running the full script
initialize(config_path="configs", version_base=None)
cfg = compose(config_name="T5start")

train_translation_dataset = load_cache_dataset(
    dataset_cfg=cfg.dataset,
    dataset_name=cfg.dataset_name,
    split="train",
) 
val_translation_dataset = load_cache_dataset(
    dataset_cfg=cfg.dataset,
    dataset_name="roszcz/maestro-v1-sustain",
    split=f"validation+test",
)

tokenizer = MultiVelocityEncoder(
    quantization_cfg=cfg.dataset.quantization,
    time_quantization_method=cfg.time_quantization_method,
)

train_dataset = MyTokenizedMidiDataset(
    dataset=train_translation_dataset,
    dataset_cfg=cfg.dataset,
    encoder=tokenizer,
)
val_dataset = MyTokenizedMidiDataset(
    dataset=val_translation_dataset,
    dataset_cfg=cfg.dataset,
    encoder=tokenizer,
)
config = T5Config(
    vocab_size=vocab_size(cfg),
    decoder_start_token_id=0,
    use_cache=False,
)
model = T5ForConditionalGeneration(config)
print("Millions of parameters:", sum(p.numel() for p in model.parameters()) / 1e6)

train_model(
    model=model,
    train_dataset=train_dataset,
    val_dataset=val_dataset,
    cfg=cfg,
)
```

### T5 
The initial training method includes using MultiTokEncoders and Absolute Time quantization.

Run training to predict start:
```bash
python train.py --config-name T5start device=cuda:0 log=True 
```
To predict velocity:
```bash
python train.py --config-name T5velocity device=cuda:0 log=True 
```
### Evaluation
Evaluation is done using average distance metric. To run evaluation use:
```bash
python -m pipelines.T5.evaluation run_name=model_run_name
```

### Dashboard
You can run velocity model prediction review dashboard by running
```bash
PYTHONPATH=. streamlit run --server.port 4466 dashboard.velocity.main
```

It works only with MidiATQuantizer for now.

### CODE REVIEW

**Comment by WojciechMat** [comment]
- [x] learning rate scheduler
- [x] T5 pipeline
- [x] distance calculation

**Comment by WojciechMat** [comment]
You can find first model trained on maestro [here](https://huggingface.co/wmatejuk/midi-T5-velocity).

**Comment by roszcz** [comment]
> You can find first model trained on maestro [here](https://huggingface.co/wmatejuk/midi-T5-velocity).

Can you add explicit instruction in the PR description on how to get a dashboard running with this checkpoint? Downloading the checkpoint from HF should also be a script (I thought we already had one for this here, but I can't find it), doesn't have to be a part of the dashboard code though 🤔  

Also, can you share more details about this model performance (loss, accuracy, training time, wandb dashboard)? Just by looking at a couple of samples, the generated velocities all look very flat compared to inputs:

image

I think we don't need the "Q.velocity" column on the dashboard. If you want to review the results as compared to the quantized values, creating charts or metrics comparing the velocity values numerically may be more insightful 🤔

File: configs/BARTpre.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  sequence_step: 2
+
+  quantization:
+    start: 20
+    duration: 3
+    velocity: 3
+
+device: "cpu"
+
+log: False
```
Comment (roszcz) [inline configs/BARTpre.yaml]:
```suggestion
log: True
```

I think we should log everything by default

File: data/dataset.py [hunk 2]
```diff
@@ -356,16 +356,14 @@ def main():
     print(piece.df)
     # ff.view.make_piano_roll_video(piece, "test.mp4")
 
-    from data.tokenizer import MultiVelocityEncoder
-
     keys = ["pitch", "start_bin", "duration_bin", "velocity_bin"]
-    encoder = MultiVelocityEncoder(cfg.quantization, keys=keys)
+    encoder = MaskedMidiEncoder(cfg.quantization, keys, 0.3)
```
Comment (roszcz) [inline data/dataset.py]: Please use named parameters, and try to avoid hard-coding hyperparameters :)

File: data/tokenizer.py [hunk 3]
```diff
@@ -237,3 +237,103 @@ def untokenize_tgt(self, tokens: list[str]) -> list[int]:
...
+        time_tokens_product = itertools.product(
+            # weird because we want to use dstart if there is a dstart_bin key, and start if there is a start_bin
+            range(self.quantization_cfg[self.keys[1][:-4]]),
+            range(self.quantization_cfg.duration),
+        )
+
+        for pitch in range(21, 109):
+            self.vocab.append(f"PITCH-{pitch}")
+        for start, duration in time_tokens_product:
+            self.vocab.append(f"TIME-{start:0.0f}-{duration:0.0f}")
```
Comment (roszcz) [inline data/tokenizer.py]: Does it make sense to try this wit separate `TIME` and `DURATION` tokens? 🤔

File: data/tokenizer.py [hunk 4]
```diff
@@ -237,3 +237,103 @@ def untokenize_tgt(self, tokens: list[str]) -> list[int]:
...
+
+    @property
+    def vocab_size(self) -> int:
+        return len(self.vocab)
+
+    def _build_vocab(self):
+        time_tokens_product = itertools.product(
+            # weird because we want to use dstart if there is a dstart_bin key, and start if there is a start_bin
+            range(self.quantization_cfg[self.keys[1][:-4]]),
+            range(self.quantization_cfg.duration),
```
Comment (roszcz) [inline data/tokenizer.py]: Please make it not weird and easy to understand :)

If this is checking if the tokenizer should use `start` or `dstart`, it should be possible to be explicit about it

Comment (WojciechMat) [reply data/tokenizer.py]: it generalizes it to both scenarios, it is possible to use dstart or start now, I can make it with ifs instead if you want :)

File: data/tokenizer.py (hunk 3 above)
Comment (WojciechMat) [reply data/tokenizer.py]: So far I think it won't matter that much so we might as well seperate them.

**Comment by WojciechMat** [comment]
You can download velocity model with 
```
python -m dashboard.download_model
```
It is true the model tends to generate very flat velocities.
Here is its [wandb run](https://wandb.ai/wmatejuk/midi-hf-transformer/runs/er9sscy7?workspace=user-wmatejuk)

File: data/tokenizer.py (hunk 4 above)
Comment (roszcz) [reply data/tokenizer.py]: I think we should be able to be more explicit about this. I also don't like the way `keys` are used in this class - we should find a better name (what type of keys are those), and make this a required argument, so get rid of the `if keys is None` part in the `__init__`.

Then getting the number of bins related to time could look like this:

```py
self.quantization_cfg[self.time_quantization_method]
```

And quantization method would ba aproperty inferred from the keys argument.

**Comment by roszcz** [comment]
> You can download velocity model with
> 
> ```
> python -m dashboard.download_model
> ```
> 
> It is true the model tends to generate very flat velocities. Here is its [wandb run](https://wandb.ai/wmatejuk/midi-hf-transformer/runs/er9sscy7?workspace=user-wmatejuk)

By looking at the wandb dashboard I wouldn't say that this model is learning anything meaningful - can you try focusing on that before setting up other pipelines?

File: data/tokenizer.py (hunk 4 above)
Comment (WojciechMat) [reply data/tokenizer.py]: done, you can check out if the way it is done now looks good enough for you

**Comment by WojciechMat** [comment]
Let's stay focused on training T5 and leave BART for future pull request.

**Comment by WojciechMat** [comment]
I've cleaned up the code. I think the main objective of this pull request was to create a training utils that fit huggingface models. The same training logic can be used for both T5 and BART, and possibly other models as well. 

It was achieved.

I think we can merge this pull request to keep the repository tidy and build experiments with different encoders and datasets on top of that.
